File: leetcode/0345. Reverse Vowels of a String/Solution.java
// https://leetcode.com/problems/reverse-vowels-of-a-string/

class Solution {
    private boolean isVowel(char c) {
        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' ||
            c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U';
    }
    
    public String reverseVowels(String s) {
        if (s.length() <= 1) {
            return s;
        }

        int i = 0;
        int j = s.length() - 1;

        char[] chars = s.toCharArray();

        while (i < j) {
            char a = chars[i];
            char b = chars[j];

            if (!isVowel(a)) {
                i++;
                continue;
            }

            if (!isVowel(b)) {
                j--;
                continue;
            }

            char t = a;
            chars[i]= b;
            chars[j]= t;

            i++;
            j--;
        }


        return String.copyValueOf(chars);
    }
}

File: leetcode/0345. Reverse Vowels of a String/Solution.py
# https://leetcode.com/problems/reverse-vowels-of-a-string/

class Solution:
    def reverseVowels(self, s: str) -> str:
        def isVowel(c: str) -> bool:
            return c in "aeiouAEIOU"
    
        n = len(s)
        if n <= 1:
            return s

        r = list(s)

        i, j = 0, n - 1
        while i < j:
            a = r[i]
            b = r[j]

            if not isVowel(a):
                i += 1
                continue

            if not isVowel(b):
                j -= 1
                continue

            r[i], r[j] = r[j], r[i]
            i += 1
            j -= 1

        return "".join(r)

s = Solution()
assert s.reverseVowels("hello") == "holle"
assert s.reverseVowels("leetcode") == "leotcede"